File: src/languages/mongolian/mongolian.ts
/**
 * Mongolian (mn) phonemizer — Standard Khalkha, canonical IPA. Cyrillic Khalkha is a DEEP
 * orthography: only the first-syllable vowel is realised full; a written non-initial SHORT vowel reduces to ə, or
 * deletes word-finally (Дани→tän, Сири→sir). Long vowels (doubled) and diphthongs stay full everywhere. On top of
 * the greedy g2p scan (g2p.ts) this module applies that reduction/deletion, final obstruent devoicing (в→f, г→k,
 * ж→t͡ʃ, з→t͡s), and final н→ŋ, then tokenizes words/numbers/punctuation.
 */
import type { Phonemizer } from "../../registry.ts";
import { assembleClauses } from "../../core/clauses.ts";
import { toSegments, type Seg } from "./g2p.ts";
import { numberToWords, spellDigits } from "./numbers.ts";
import { bichigToCyrillic, isBichig } from "./mongolBichig.ts";
import { normalizeMongolian } from "./normalize.ts";
import { MANIFEST } from "./manifest.ts";

// Word-final obstruent devoicing: only в→f (Женев→dʒɛnɛf). Final г stays voiced (хаг→xaɡ, хурга→xʊrəɢ).
const DEVOICE: Record<string, string> = { w: "f" };
// A non-initial SHORT vowel reduces to a quality that keeps the ORIGINAL letter's rounding (судар а→ə, but Орос о→ʊ,
// Төвөд ө→ɵ). Unround → ə, back-round → ʊ, front-round → ɵ/u.
const REDUCED_OF: Record<string, string> = { a: "ə", e: "ə", i: "ə", æ: "ə", ɔ: "ʊ", ʊ: "ʊ", o: "ʊ", ɵ: "ɵ", œ: "ɵ", u: "u" };
const isCons = (s: Seg | undefined): boolean => s !== undefined && !s.nucleus;

/** Apply the deep-orthography vowel reduction: keep the first nucleus full; a non-initial non-final SHORT vowel
 *  reduces (quality by the original letter's rounding); a word-final short vowel DELETES (Дани→tän, хаана→xaːn). If
 *  deleting a final vowel leaves a final consonant CLUSTER, an epenthetic reduced vowel breaks it (хурга→xʊrəɢ). Long
 *  vowels / diphthongs stay full. */
function reduce(segs: Seg[], initialFull = true, keepNonFinal = false): Seg[] {
    let seen = !initialFull; // a bound morpheme (a -suffix entry) has NO full first vowel — everything reduces
    const out: Seg[] = [];
    let epenthesis: string | null = null; // reduced quality of a deleted final vowel, to break a resulting cluster
    for (let i = 0; i < segs.length; i++) {
        const s = segs[i]!;
        if (!s.nucleus) { out.push(s); continue; }
        if (!seen) { seen = true; out.push(s); continue; } // first vowel is full
        if (!s.short) { out.push(s); continue; } // long vowel / diphthong stays full
        const reduced = REDUCED_OF[s.ph] ?? "ə";
        if (i === segs.length - 1) { epenthesis = reduced; continue; } // word-final short vowel → delete
        // A mixed-harmony (loanword) word keeps its non-final vowels FULL — reduction is a native-word process.
        out.push(keepNonFinal ? s : { ph: reduced, nucleus: true, short: true });
    }
    // Epenthesis: a deleted final vowel that leaves a final consonant cluster surfaces as a reduced vowel inside it.
    if (epenthesis && out.length >= 2 && isCons(out[out.length - 1]) && isCons(out[out.length - 2])) {
        out.splice(out.length - 1, 0, { ph: epenthesis, nucleus: true, short: true });
    }
    return out;
}

/**
 * ⟨ї⟩ U+0457 IS ⟨ү⟩ — A LEGACY-CODEPAGE ARTEFACT, not a Ukrainian letter that wandered in.
 *
 * The pre-Unicode Mongolian Cyrillic fonts (and the Windows-1251 remaps that came with them) had no slot for
 * ⟨ү⟩ or ⟨ө⟩ and borrowed the Ukrainian/Church-Slavonic codepoints for them; text typed that way and pasted
 * into the wiki keeps the wrong codepoint. The corpus reads unambiguously: ⟨ї⟩ ×8 against ⟨ү⟩ ×2,703, and
 * every one of the eight is a word that exists with ⟨ү⟩ — `бїр` (бүр), `бїлэг` (бүлэг), `їр` (үр). It is
 * outside the letter tables, so the words read `pr`, `pɮeɡ`, `r` — a vowelless consonant string.
 *
 * ⟨ѳ⟩ U+0473 (fita) → ⟨ө⟩ is the OTHER HALF OF THE SAME PAIR, and it is ×0 in this artifact — it is included
 * because Bashkir's artifact carries exactly that substitution (⟨ѳ⟩ ×7 against ⟨ө⟩ ×1,323, `кѳньяғында`,
 * `һѳрѳлгән`) from the same font generation, and because fita is in no modern Cyrillic alphabet at all, so
 * there is no reading it could be taken from. Stated rather than left silent: unattested HERE, today.
 *
 * ⚠ NOT A REGISTRY-LEVEL CONFUSABLE FOLD. `foldCyrillicConfusables` maps LATIN look-alikes into Cyrillic; ⟨ї⟩
 * is a real Cyrillic letter of another language's alphabet, and folding it fleet-wide would corrupt Ukrainian,
 * where it is the letter ⟨ї⟩ /ji/. The claim is about MONGOLIAN text specifically, so it lives here.
 */
const LEGACY_CODEPAGE: Readonly<Record<string, string>> = { "ї": "ү", "ѳ": "ө" };
const LEGACY_RE = new RegExp(`[${Object.keys(LEGACY_CODEPAGE).join("")}]`, "gu");
const foldLegacyCodepage = (w: string): string => w.replace(LEGACY_RE, (c) => LEGACY_CODEPAGE[c]!);

/** One Mongolian word → canonical IPA. A word in the traditional Mongolian script (Mongol bichig) is transliterated
 *  to Cyrillic first (mongolBichig.ts) and then run through the same pipeline. */
export function phonemizeWord(word: string): string {
    if (isBichig(word)) word = bichigToCyrillic(word);
    const w = foldLegacyCodepage(word.toLowerCase());
    const raw = toSegments(w);
    if (!raw.some((s) => s.nucleus)) return raw.map((s) => s.ph).join(""); // vowelless (a letter name) — no rules
    // Final н → ŋ keyed on the ORIGINAL spelling (before vowel deletion): н that is word-final or before a velar/
    // uvular → ŋ (заан→tsaːŋ), but н still followed by a (to-be-deleted) vowel stays n (хаана→xaːn).
    for (let i = 0; i < raw.length; i++) {
        if (raw[i]!.ph !== "n") continue;
        const next = raw[i + 1]?.ph;
        if (next === undefined || /^[ɡɢkxχq]/u.test(next)) raw[i]!.ph = "ŋ";
    }
    const bound = /^[-­]/u.test(word); // a -suffix entry (hyphen / soft-hyphen): its leading vowel is non-initial
    // A word mixing BACK (а/о/у/я/ё/ю) and FRONT (э/ө/ү/е) vowels violates Mongolian vowel harmony → it is a loanword
    // (Герман, Австри), whose non-initial vowels stay FULL rather than reducing. (и/ы are neutral.)
    const loan = /[аоуяёю]/u.test(w) && /[эөүе]/u.test(w);
    let segs = reduce(raw, !bound, loan);
    // Final obstruent devoicing on the last segment — but NOT when the written word ends in ь (the soft sign was
    // dropped, so the obstruent isn't truly final: Говь → ɢow, not ɢof).
    const last = segs[segs.length - 1];
    if (last && DEVOICE[last.ph] && !w.endsWith("ь")) last.ph = DEVOICE[last.ph]!;
    return segs.map((s) => s.ph).join("");
}

const CLAUSE_MARK = MANIFEST.clausePunctuation;
const TOKEN = /([Ѐ-ӿᠠ-ᡂ᠋-᠎‌‍ ]+)|(\d+)|([.!?…,;:])/gu;

class MongolianPhonemizer implements Phonemizer {
    text(input: string): string {
        // NORMALIZATION FIRST — see normalize.ts. Every unit, sign, ordinal and initialism reading lives
        // there rather than in `core/normalizeSymbols.ts`: Mongolian glues a CASE SUFFIX to the percent sign
        // (`67%-ийг` → *хувийг*), which is trap 14 and which the tier's invariant strings cannot express.
        return assembleClauses(normalizeMongolian(input), TOKEN, (m, sink) => {
            if (m[1]) sink.emit(phonemizeWord(m[1]));
            // cardinal → words → IPA. ⚠ ABOVE 2^53 `numberToWords` RETURNS "" AND THE NUMBER USED TO VANISH:
            // the split produced one empty word and nothing was emitted, so the sentence scanned without its
            // numeral. Refusing to compose is right (the float has lost the low digits); the else was missing.
            else if (m[2]) {
                const composed = numberToWords(Number(m[2]));
                for (const wd of (composed === "" ? spellDigits(m[2]) : composed).split(" "))
                    sink.emit(phonemizeWord(wd));
            }
            else if (m[3]) { const mk = CLAUSE_MARK[m[3]]; if (mk) sink.pause(mk); }
        });
    }
}

/** Build the Mongolian phonemizer (greedy Cyrillic g2p + deep-orthography reduction + final devoicing). */
export function createMongolian(): Phonemizer {
    return new MongolianPhonemizer();
}

File: csharp/Vernacula.Phonemizer/Languages/Mongolian/Mongolian.cs
/**
 * Mongolian (mn) phonemizer — Standard Khalkha. Cyrillic Khalkha is a DEEP orthography: only the first-syllable
 * vowel is realised full, a non-initial short vowel reduces to ə and a word-final one deletes. On top of the
 * greedy g2p scan this module applies that reduction/deletion, the ⟨в⟩ final devoicing and final н→ŋ, then
 * tokenizes. Ported from src/languages/mongolian/mongolian.ts — see that file for the corpus evidence,
 * including the ⟨ї⟩/⟨ѳ⟩ legacy-codepage argument.
 */
using Vernacula.Phonemizer.Core;

namespace Vernacula.Phonemizer.Languages.Mongolian;

public static class MongolianPhonemizer
{
    /** Word-final obstruent devoicing: only в→f. Final г stays voiced. */
    private static readonly Dictionary<string, string> DEVOICE = new() { ["w"] = "f" };

    /** A non-initial SHORT vowel reduces to a quality that keeps the ORIGINAL letter's rounding. */
    private static readonly Dictionary<string, string> REDUCED_OF = new()
    {
        ["a"] = "ə", ["e"] = "ə", ["i"] = "ə", ["æ"] = "ə", ["ɔ"] = "ʊ", ["ʊ"] = "ʊ",
        ["o"] = "ʊ", ["ɵ"] = "ɵ", ["œ"] = "ɵ", ["u"] = "u",
    };

    private static bool IsCons(Seg? s) => s is not null && !s.Nucleus;

    /** Apply the deep-orthography vowel reduction: the first nucleus stays full; a non-initial non-final SHORT
     *  vowel reduces; a word-final short vowel DELETES, and an epenthetic reduced vowel breaks the consonant
     *  cluster that leaves. Long vowels / diphthongs stay full. */
    private static List<Seg> Reduce(List<Seg> segs, bool initialFull = true, bool keepNonFinal = false)
    {
        var seen = !initialFull; // a bound morpheme has NO full first vowel — everything reduces
        var outp = new List<Seg>();
        string? epenthesis = null;
        for (var i = 0; i < segs.Count; i++)
        {
            var s = segs[i];
            if (!s.Nucleus) { outp.Add(s); continue; }
            if (!seen) { seen = true; outp.Add(s); continue; } // first vowel is full
            if (!s.Short) { outp.Add(s); continue; } // long vowel / diphthong stays full
            var reduced = REDUCED_OF.GetValueOrDefault(s.Ph, "ə");
            if (i == segs.Count - 1) { epenthesis = reduced; continue; } // word-final short vowel → delete
            // A mixed-harmony (loanword) word keeps its non-final vowels FULL.
            outp.Add(keepNonFinal ? s : new Seg { Ph = reduced, Nucleus = true, Short = true });
        }
        if (epenthesis is not null && outp.Count >= 2 && IsCons(outp[^1]) && IsCons(outp[^2]))
            outp.Insert(outp.Count - 1, new Seg { Ph = epenthesis, Nucleus = true, Short = true });
        return outp;
    }

    /** ⟨ї⟩ U+0457 IS ⟨ү⟩ and ⟨ѳ⟩ U+0473 is ⟨ө⟩ — a legacy-codepage artefact of the pre-Unicode Mongolian
     *  fonts, not another language's letters wandering in. ⚠ NOT a registry-level confusable fold: ⟨ї⟩ is a
     *  real Ukrainian letter, so the claim is about MONGOLIAN text specifically and lives here. */
    private static readonly Dictionary<string, string> LEGACY_CODEPAGE = new() { ["ї"] = "ү", ["ѳ"] = "ө" };
    private static readonly JsRe LEGACY_RE =
        JsRegex.Compile($"[{string.Concat(LEGACY_CODEPAGE.Keys)}]", "gu");
    private static string FoldLegacyCodepage(string w) =>
        LEGACY_RE.Replace(w, m => LEGACY_CODEPAGE[m.Value]);

    private static readonly JsRe VELAR_AFTER_N = JsRegex.Compile("^[ɡɢkxχq]", "u");
    private static readonly JsRe BOUND_PREFIX = JsRegex.Compile("^[-­]", "u");
    private static readonly JsRe HAS_BACK = JsRegex.Compile("[аоуяёю]", "u");
    private static readonly JsRe HAS_FRONT = JsRegex.Compile("[эөүе]", "u");

    /** One Mongolian word → canonical IPA. A Mongol-bichig word is transliterated to Cyrillic first. */
    public static string PhonemizeWord(string word)
    {
        if (MongolBichig.IsBichig(word)) word = MongolBichig.BichigToCyrillic(word);
        var w = FoldLegacyCodepage(Js.ToLowerCase(word));
        var raw = G2p.ToSegments(w);
        if (!raw.Any(s => s.Nucleus)) return string.Concat(raw.Select(s => s.Ph)); // vowelless (a letter name)
        // Final н → ŋ keyed on the ORIGINAL spelling (before vowel deletion).
        for (var i = 0; i < raw.Count; i++)
        {
            if (raw[i].Ph != "n") continue;
            var next = i + 1 < raw.Count ? raw[i + 1].Ph : null;
            if (next is null || VELAR_AFTER_N.IsMatch(next)) raw[i].Ph = "ŋ";
        }
        var bound = BOUND_PREFIX.IsMatch(word); // a -suffix entry: its leading vowel is non-initial
        // A word mixing BACK and FRONT vowels violates harmony → a loanword, whose non-initial vowels stay FULL.
        var loan = HAS_BACK.IsMatch(w) && HAS_FRONT.IsMatch(w);
        var segs = Reduce(raw, !bound, loan);
        // Final devoicing — but NOT when the written word ends in ь (the obstruent isn't truly final).
        var last = segs.Count > 0 ? segs[^1] : null;
        if (last is not null && DEVOICE.TryGetValue(last.Ph, out var dv) && !w.EndsWith("ь", StringComparison.Ordinal))
            last.Ph = dv;
        return string.Concat(segs.Select(s => s.Ph));
    }

    private static IReadOnlyDictionary<string, string> CLAUSE_MARK => Manifest.MANIFEST.ClausePunctuation;
    private static readonly JsRe TOKEN = JsRegex.Compile("([Ѐ-ӿᠠ-ᡂ᠋-᠎‌‍\\u202f]+)|(\\d+)|([.!?…,;:])", "gu");

    private sealed class Engine : ILanguage
    {
        public string Text(string input)
        {
            // NORMALIZATION FIRST — every unit, sign, ordinal and initialism reading lives in Normalize.cs
            // rather than in the shared symbol tier; see the TS header for why the tier cannot say it.
            return Clauses.AssembleClauses(Normalize.NormalizeMongolian(input), TOKEN, (m, sink) =>
            {
                if (m.Groups[1].Success && m.Groups[1].Value.Length > 0)
                    sink.Emit(PhonemizeWord(m.Groups[1].Value));
                // ⚠ ABOVE 2^53 `NumberToWords` RETURNS "" and the number used to VANISH — the else is the fix.
                else if (m.Groups[2].Success && m.Groups[2].Value.Length > 0)
                {
                    var tok = m.Groups[2].Value;
                    var composed = Numbers.NumberToWords(Js.Number(tok));
                    foreach (var wd in (composed == "" ? Numbers.SpellDigits(tok) : composed).Split(' '))
                        sink.Emit(PhonemizeWord(wd));
                }
                else if (m.Groups[3].Success && m.Groups[3].Value.Length > 0)
                {
                    if (CLAUSE_MARK.TryGetValue(m.Groups[3].Value, out var mk) && mk.Length > 0) sink.Pause(mk);
                }
            });
        }
    }

    /** Build the Mongolian phonemizer (greedy Cyrillic g2p + deep-orthography reduction + final devoicing). */
    public static ILanguage CreateMongolian() => new Engine();

    internal static void RegisterSelf() => Registry.Register("mongolian", () => CreateMongolian());
}
